ThirdPartyAdapters/chartboost/chartboost/src/main/java/com/google/ads/mediation/chartboost/ChartboostAdapter.java

```diff
@@ -186,6 +186,101 @@ public void onPause() {
+    }
+
+    @Override
+    public View getBannerView() {
+        WeakReference<ChartboostBanner> ref = ChartboostSingleton.getChartboostBanner(mChartboostParams.getLocation());
+        if(ref != null) {
+            return ref.get();
+        }
+        return null;
+    }
+
+    private ChartboostBannerListener createChartboostBannerListener() {
+        return new ChartboostBannerListener() {
+            @Override
+            public void onAdCached(ChartboostCacheEvent chartboostCacheEvent, ChartboostCacheError chartboostCacheError) {
+                if(mMediationBannerListener!=null) {
+                    if(chartboostCacheError == null) {
+                        mMediationBannerListener.onAdLoaded(ChartboostAdapter.this);
+                    } else {
+                        reportChartboostErrorToAdMob(chartboostCacheError.code);
+                    }
+                }
+            }
+
+            @Override
+            public void onAdShown(ChartboostShowEvent chartboostShowEvent, ChartboostShowError chartboostShowError) {
+                if(mMediationBannerListener!=null) {
+                    if (chartboostShowError == null) {
+                        mMediationBannerListener.onAdOpened(ChartboostAdapter.this);
+                    } else {
+                        reportChartboostErrorToAdMob(chartboostShowError.code);
+                    }
+                }
+            }
+
+            @Override
+            public void onAdClicked(ChartboostClickEvent chartboostClickEvent, ChartboostClickError chartboostClickError) {
+                if (mMediationBannerListener != null) {
+                    if (chartboostClickError == null) {
+                        mMediationBannerListener.onAdClicked(ChartboostAdapter.this);
+                        mMediationBannerListener.onAdLeftApplication(ChartboostAdapter.this);
+                    } else {
+                        reportChartboostErrorToAdMob(chartboostClickError.code);
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * Parse and report chartboost error code to AdMob error
+     * @param code
+     */
+    private void reportChartboostErrorToAdMob(int code) {
+        switch (code) {
+            case 1: mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this, AdRequest.ERROR_CODE_NETWORK_ERROR);
+            case 5: mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this, AdRequest.ERROR_CODE_NETWORK_ERROR);
+            case 6: mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this, AdRequest.ERROR_CODE_NO_FILL);
+            case 7: mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this, AdRequest.ERROR_CODE_NETWORK_ERROR);
+            case 16: mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this, AdRequest.ERROR_CODE_INVALID_REQUEST);
+            default: mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this, AdRequest.ERROR_CODE_INTERNAL_ERROR);
+        }
```

Comment:
Actually we could make the function just do the error mapping, and call mMediationBannerListener.onAdFailedToLoad whenever it is used. 
It would read nicer since the counterpart happy path calls (mMediationBannerListener.onAdLoaded, mMediationBannerListener.onAdOpened) are called there, so we would keep the same level of abstraction.